### ISSUE TITLE
no-use-before-define edge case

### ISSUE THREAD

**Issue by albertosantini** [eslint/eslint#4533]
I get the following error

`4:5  error  "baz" was used before it was defined  no-use-before-define`

Notice the same function commented. 
That version doesn't raise the error.

Do we consider it a legit case of `no-use-before-define` or an edge case "won-t fix" or anything different?

The real case is json_parse function in https://github.com/douglascrockford/JSON-js/blob/master/json_parse.js.

With `eslint 1.10.1` and with the following `.eslintrc.yml` and `eslint-config-eslint@1.0.1` plugin:
```
root: true

env:
    node: true

extends: "eslint"

rules:
    new-cap: [2, {capIsNew: false}]
    no-use-before-define: [2, "nofunc"]
    require-jsdoc: 0
    space-before-function-paren: [2, {"anonymous": "always", "named": "never"}]
    spaced-comment: [0, "always", { exceptions: ["-"]}]
```

and the following snippet:

```
"use strict";

function bar() {
    baz();
}

var baz = (function () {
}());

// function baz() {
// }

console.log(bar);
```

**Comment by eslintbot** [comment]
@albertosantini Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by platinumazure** [comment]
Function declarations (your commented version) are hoisted. Function
expressions are not, but the variable declaration is. Your uncommented
version is equivalent to the following:

```js
var myFunc;

var callingFunc = function () {
    myFunc();
};

myFunc = function () { };
```

The rule is stylistic and meant to catch cases where a variable use comes
before its definition and the rule is working as designed.
